pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -1158,10 +863,11 @@ func (pl *dynamicResources) Filter(ctx context.Context, cs *framework.CycleState
 		logger.V(10).Info("filtering based on resource claims of the pod", "pod", klog.KObj(pod), "node", klog.KObj(node), "resourceclaim", klog.KObj(claim))
 
 		if claim.Status.Allocation != nil {
-			if nodeSelector := state.informationsForClaim[index].availableOnNode; nodeSelector != nil {
+			for _, nodeSelector := range state.informationsForClaim[index].availableOnNodes {
 				if !nodeSelector.Match(node) {
 					logger.V(5).Info("AvailableOnNodes does not match", "pod", klog.KObj(pod), "node", klog.KObj(node), "resourceclaim", klog.KObj(claim))
 					unavailableClaims = append(unavailableClaims, index)
+					break
```

Comment:
It's the same logic as before, just adapted to the new API in 1.31: instead of one ResourceClass per ResourceClaim (and thus one "available on nodes" selector), there's now one such selector per DeviceClass and thus per request inside the claim. Therefore the old `if` check got replaced with a `for` loop.

The rest of the code below is also similar, but one level of indention got removed:
- with the `continue` in the "already allocated" case here
- with the `return` if "deallocation requested"
- by removing the special case for "immediate allocation"

Reply:
Another change that affects the flow is that all claims that need to be allocated via structured parameters get collected and then allocated together.

Reply:
> Therefore the old if check got replaced with a for loop.

makes sense, i overlooked the if -> for change.